### PR TITLE
use composite keys in relation reference widget

### PR DESCRIPTION
**Pull request by 3nids**

### CODE REVIEW

File: src/gui/editorwidgets/qgsrelationreferencewidget.cpp [hunk 1]
```diff
@@ -45,6 +45,20 @@
 #include "qgsidentifymenu.h"
 
 
+bool qVariantListIsNull( const QVariantList &list )
+{
+  if ( list.isEmpty() )
+    return false;
+
+  for ( int i = 0; i < list.size(); ++i )
```
Comment (m-kuhn) [inline src/gui/editorwidgets/qgsrelationreferencewidget.cpp]: Can this be a for loop?

File: src/gui/editorwidgets/qgsrelationreferencewidget.cpp [hunk 2]
```diff
@@ -271,8 +288,14 @@ void QgsRelationReferenceWidget::setForeignKey( const QVariant &value )
   {
     // Attributes from the referencing layer
     QgsAttributes attrs = QgsAttributes( mReferencingLayer->fields().count() );
-    // Set the value on the foreign key field of the referencing record
-    attrs[ mReferencingLayer->fields().lookupField( mRelation.fieldPairs().at( 0 ).first )] = value;
+    // Set the value on the foreign key fields of the referencing record
+
+    const QList<QgsRelation::FieldPair> fieldPairs = mRelation.fieldPairs();
+    for ( int i = 0; i < fieldPairs.count(); i++ )
```
Comment (m-kuhn) [inline src/gui/editorwidgets/qgsrelationreferencewidget.cpp]: Better save than sorry

```suggestion
    int fieldCount = std::min( fieldPairs.count(), values.count() );
    for ( int i = 0; i < fieldCount; i++ )
```

File: src/gui/editorwidgets/qgsrelationreferencewidget.cpp [hunk 3]
```diff
@@ -283,23 +306,26 @@ void QgsRelationReferenceWidget::setForeignKey( const QVariant &value )
       return;
     }
 
-    mForeignKey = mFeature.attribute( mReferencedFieldIdx );
+    mForeignKeys.clear();
+    for ( const QString &fieldName : mReferencedFields )
```
Comment (m-kuhn) [inline src/gui/editorwidgets/qgsrelationreferencewidget.cpp]:
```suggestion
    for ( const QString &fieldName : qgis::as_const( mReferencedFields ) )
```

File: src/gui/editorwidgets/qgsrelationreferencewidget.cpp [hunk 4]
```diff
@@ -283,23 +306,26 @@ void QgsRelationReferenceWidget::setForeignKey( const QVariant &value )
...
 
     QgsExpression expr( mReferencedLayer->displayExpression() );
     QgsExpressionContext context( QgsExpressionContextUtils::globalProjectLayerScopes( mReferencedLayer ) );
     context.setFeature( mFeature );
     QString title = expr.evaluate( &context ).toString();
     if ( expr.hasEvalError() )
     {
-      title = mFeature.attribute( mReferencedFieldIdx ).toString();
+      QStringList titleFields;
+      for ( const QString &fieldName : mReferencedFields )
```
Comment (m-kuhn) [inline src/gui/editorwidgets/qgsrelationreferencewidget.cpp]: as_const

File: src/gui/editorwidgets/qgsrelationreferencewidget.cpp [hunk 5]
```diff
@@ -346,7 +372,13 @@ void QgsRelationReferenceWidget::deleteForeignKey()
       nullText = tr( "%1 (no selection)" ).arg( nullValue );
     }
     mLineEdit->setText( nullText );
-    mForeignKey = QVariant( QVariant::Int );
+    QVariantList nullAttributes;
+    for ( const QString &fieldName : mReferencedFields )
```
Comment (m-kuhn) [inline src/gui/editorwidgets/qgsrelationreferencewidget.cpp]: as_const

File: src/gui/editorwidgets/qgsrelationreferencewidget.cpp [hunk 6]
```diff
@@ -742,22 +783,27 @@ void QgsRelationReferenceWidget::featureIdentified( const QgsFeature &feature )
     QString title = expr.evaluate( &context ).toString();
     if ( expr.hasEvalError() )
     {
-      title = feature.attribute( mReferencedFieldIdx ).toString();
+      QStringList titleFields;
+      for ( const QString &fieldName : mReferencedFields )
```
Comment (m-kuhn) [inline src/gui/editorwidgets/qgsrelationreferencewidget.cpp]: as_const

File: src/gui/editorwidgets/qgsrelationreferencewidget.cpp [hunk 7]
```diff
@@ -742,22 +783,27 @@ void QgsRelationReferenceWidget::featureIdentified( const QgsFeature &feature )
...
+        titleFields << mFeature.attribute( fieldName ).toString();
+      title = titleFields.join( QStringLiteral( " " ) );
     }
     mLineEdit->setText( title );
-    mForeignKey = feature.attribute( mReferencedFieldIdx );
+    mForeignKeys.clear();
+    for ( const QString &fieldName : mReferencedFields )
```
Comment (m-kuhn) [inline src/gui/editorwidgets/qgsrelationreferencewidget.cpp]: as_const

File: src/gui/editorwidgets/qgsrelationreferencewidget.cpp [hunk 8]
```diff
@@ -909,9 +955,12 @@ void QgsRelationReferenceWidget::addEntry()
 
   if ( mEditorContext.vectorLayerTools()->addFeature( mReferencedLayer, attributes, QgsGeometry(), &f ) )
   {
-    Q_NOWARN_DEPRECATED_PUSH
-    mComboBox->setIdentifierValue( f.attribute( mReferencedFieldIdx ) );
-    Q_NOWARN_DEPRECATED_POP
+    QVariantList attrs;
+    for ( const QString &fieldName : mReferencedFields )
```
Comment (m-kuhn) [inline src/gui/editorwidgets/qgsrelationreferencewidget.cpp]: as_const

File: src/gui/editorwidgets/qgsrelationreferencewidgetwrapper.cpp [hunk 9]
```diff
@@ -124,13 +131,70 @@ void QgsRelationReferenceWidgetWrapper::showIndeterminateState()
...
+    {
+      values << QVariant();
+    }
+    return values;
+  }
+  else
+  {
+    QVariantList values = mWidget->foreignKeys();
+    const QList<QgsRelation::FieldPair> fieldPairs = mWidget->relation().fieldPairs();
+    for ( int i = 0; i < fieldPairs.count(); i++ )
```
Comment (m-kuhn) [inline src/gui/editorwidgets/qgsrelationreferencewidgetwrapper.cpp]:
```suggestion
    int fieldCount = std::min( fieldPairs.count(), values.count() );
    for ( int i = 0; i < fieldCount; i++ )
```

File: src/gui/editorwidgets/qgsrelationreferencewidget.cpp (hunk 5 above)
Comment (3nids) [reply src/gui/editorwidgets/qgsrelationreferencewidget.cpp]:
```suggestion
    for ( const QString &fieldName : qgis::as_const( mReferencedFields ) )
```

File: src/gui/editorwidgets/qgsrelationreferencewidget.cpp (hunk 7 above)
Comment (3nids) [reply src/gui/editorwidgets/qgsrelationreferencewidget.cpp]:
```suggestion
    for ( const QString &fieldName : qgis::as_const( mReferencedFields ) )
```

File: src/gui/editorwidgets/qgsrelationreferencewidget.cpp (hunk 8 above)
Comment (3nids) [reply src/gui/editorwidgets/qgsrelationreferencewidget.cpp]:
```suggestion
    for ( const QString &fieldName : qgis::as_const( mReferencedFields ) )
```

File: src/gui/editorwidgets/qgsrelationreferencewidget.cpp (hunk 4 above)
Comment (3nids) [reply src/gui/editorwidgets/qgsrelationreferencewidget.cpp]:
```suggestion
      for ( const QString &fieldName : qgis::as_const( mReferencedFields ) )
```

File: src/gui/editorwidgets/qgsrelationreferencewidget.cpp (hunk 6 above)
Comment (3nids) [reply src/gui/editorwidgets/qgsrelationreferencewidget.cpp]:
```suggestion
      for ( const QString &fieldName : qgis::as_const( mReferencedFields ) )
```

**Pull request merged**